subprojects/instant-execution/src/integTest/groovy/org/gradle/instantexecution/InstantExecutionUnsupportedTypesIntegrationTest.groovy

```diff
@@ -0,0 +1,174 @@
+import org.gradle.api.internal.attributes.DefaultDisambiguationRuleChain
+import org.gradle.api.internal.project.DefaultProject
+import org.gradle.api.internal.tasks.DefaultSourceSet
+import org.gradle.api.internal.tasks.DefaultSourceSetContainer
+import org.gradle.api.internal.tasks.DefaultTaskContainer
+import org.gradle.api.invocation.Gradle
+import org.gradle.api.tasks.SourceSet
+import org.gradle.api.tasks.SourceSetContainer
+import org.gradle.api.tasks.TaskContainer
+import org.gradle.groovy.scripts.internal.DefaultScriptCompilationHandler.ScriptClassLoader
+import org.gradle.initialization.DefaultSettings
+import org.gradle.internal.locking.DefaultDependencyLockingHandler
+import org.gradle.invocation.DefaultGradle
+import spock.lang.Unroll
+
+import java.util.concurrent.Executor
+import java.util.concurrent.Executors.DefaultThreadFactory
+import java.util.concurrent.Executors.FinalizableDelegatedExecutorService
+import java.util.concurrent.ThreadFactory
+
+
+class InstantExecutionUnsupportedTypesIntegrationTest extends AbstractInstantExecutionIntegrationTest {
+
+    @Unroll
+    def "warns when task field references an object of type #baseType"() {
+        buildFile << """
+            plugins { id "java" }
+
+            class SomeBean {
+                private ${baseType} badReference
+            }
+
+            class SomeTask extends DefaultTask {
+                private final ${baseType} badReference
+                private final bean = new SomeBean()
+
+                SomeTask() {
+                    badReference = ${reference}
+                    bean.badReference = ${reference}
+                }
+
+                @TaskAction
+                void run() {
+                    println "this.reference = " + badReference
+                    println "bean.reference = " + bean.badReference
+                }
+            }
+
+            task other
+            task broken(type: SomeTask)
+        """
+
+        when:
+        instantRun "broken"
+
+        then:
+        outputContains("""
+            2 instant execution problems were found, 2 of which seem unique:
+              - field 'badReference' from type 'SomeTask': cannot serialize object of type '${concreteType}', a subtype of '${baseType}', as these are not supported with instant execution.
+              - field 'badReference' from type 'SomeBean': cannot serialize object of type '${concreteType}', a subtype of '${baseType}', as these are not supported with instant execution.
+            See the complete report at
+        """.stripIndent())
+
+        when:
+        instantRun "broken"
+
+        then:
+        outputContains("this.reference = null")
+        outputContains("bean.reference = null")
+
+        where:
+        concreteType                               | baseType                            | reference
+        // Live JVM state
+        ScriptClassLoader.name                     | ClassLoader.name                    | "getClass().classLoader"
```

Comment:
⭕️ I would remove all the `.name` accessors from the `where` table and add two variables after the table to hold thee names.